File: beecrowd/1112 - Schweisen.rs
#![allow(unused_imports)]
#![allow(dead_code)]
use std::cmp::*;
use std::collections::*;

struct Scanner {
   buffer : VecDeque<String>,
   eof: bool
}
 
impl Scanner {
 
   fn new() -> Scanner {
      Scanner {
         buffer: VecDeque::new(),
         eof: false
      }
   }
   
   fn next<T : std::str::FromStr + Default >(&mut self) -> T {
 
      if self.buffer.len() == 0 {
         let mut input = String::new();
         
         match std::io::stdin().read_line(&mut input) {
            Ok(0) => self.eof = true,
            Ok(_) => {},
            Err(e) => panic!("panico {}", e)
         };

         for word in input.split_whitespace() {
            self.buffer.push_back(word.to_string())
         }
      }

      if self.eof {
         return T::default();
      }
 
      let front = self.buffer.pop_front().unwrap();
      
      front.parse::<T>().ok().unwrap()
   }

   fn read_line(&mut self) -> String {
      let mut input = String::new();
         
      match std::io::stdin().read_line(&mut input) {
         Ok(0) => self.eof = true,
         Ok(_) => {},
         Err(e) => panic!("{}", e)
      };

      if input.ends_with('\n') {
         input.pop();
         if input.ends_with('\r') {
            input.pop();
         }
      }

      return input;
   }

   fn is_eof(&self) -> bool {
      self.eof
   }
}

struct BIT2D {
    data: Vec<Vec<i32>>,
    rows: usize,
    columns: usize
}

impl BIT2D {

    fn new(rows: usize, columns: usize) -> BIT2D {

        BIT2D { 
            data: vec![vec![0; columns + 1]; rows + 1], 
            rows: rows, 
            columns: columns 
        }
    }

    fn query(&self, mut x: i32, mut y: i32) -> i32 {
        x += 1;
        y += 1;

        if x < 0 || y < 0 {
            panic!("somente indices não negativos são aceitos");
        }

        let mut answer = i32::default();

        while x > 0 {
            let mut k = y;

            while k > 0 {
                answer += self.data[x as usize][k as usize];
                k -= k & -k;
            }

            x -= x & -x;
        }

        return answer;
    }

    fn calculate(&self, x0: i32, y0: i32, x1: i32, y1: i32) -> i32 {
        return self.query(x1, y1) - self.query(x1, y0 - 1) - self.query(x0 - 1, y1) + self.query(x0 - 1, y0 - 1);
    }

    fn update(&mut self, mut x: i32, mut y: i32, value: i32) {
        x += 1;
        y += 1;

        if x < 0 || y < 0 {
            panic!("somente indices não negativos são aceitos");
        }

        while x as usize  <= self.rows {

            let mut k = y;

            while k as usize <= self.columns {
                self.data[x as usize][k as usize] += value;
                k += k & -k;
            }

            x += x & -x;
        }
    }

}

fn main() {
    let mut sc = Scanner::new();

    loop {

        let x = sc.next::<usize>();
        let y = sc.next::<usize>();
        let p = sc.next::<i32>();

        if x == 0 && y == 0 && p == 0 {
            break;
        }

        let mut bit = BIT2D::new(x, y);

        let q = sc.next::<usize>();

        for _i in 0..q {

            let t = sc.next::<char>();

            if t == 'A' {
                let n = sc.next::<i32>();
                let x = sc.next::<i32>();
                let y = sc.next::<i32>();

                bit.update(x, y, n);
            } else {
                let mut x0 = sc.next::<i32>();
                let mut y0 = sc.next::<i32>();
                let mut x1 = sc.next::<i32>();
                let mut y1 = sc.next::<i32>();

                if x0 > x1 {
                    std::mem::swap(&mut x0, &mut x1);
                }

                if y0 > y1 {
                    std::mem::swap(&mut y0, &mut y1);
                }

                println!("{}", p * bit.calculate(x0, y0, x1, y1));
            }
        }

        print!("\n");
    }
}

File: beecrowd/1112 - Schweisen.cpp
#include <bits/stdc++.h>

using namespace std;

class BIT2D 
{

public:

	BIT2D(int rows, int columns): 
		_data {vector<vector<int>>(rows + 1, vector<int>(columns + 1, 0))},
		_rows {rows},
		_columns {columns}
	{
		
	}

	void update(int x, int y, int value)
	{
		++x;
		++y;

		for(; x <= _rows; x += x & -x)
			for(int k = y; k <= _columns; k += k & -k)
				_data[x][k] += value;
	}

	int query(int x, int y)
	{
		int answer {};

		++x;
		++y;

		for(; x > 0; x -= x & -x)
			for(int k = y; k > 0; k -= k & -k)
				answer += _data[x][k];

		return answer;
	}

	int query(int x0, int y0, int x1, int y1)
	{
		return query(x1, y1) - query(x1, y0 - 1) - query(x0 - 1, y1) + query(x0 - 1, y0 - 1);
	}	

private:
	vector<vector<int>> _data;
	int _rows;
	int _columns;
};

int main() 
{
	ios_base::sync_with_stdio(false);
	cin.tie(0);

	while(true)
	{
		int rows, columns, p, queries;

		cin >> rows >> columns >> p;

		if(rows + columns + p == 0)
			break;

		cin >> queries;

		BIT2D bit(rows, columns);

		while(queries--)
		{
			char type;

			cin >> type;

			if(type == 'A')
			{
				int n, x, y;
		
				cin >> n >> x >> y;

				bit.update(x, y, n);
			}
			else
			{
				int x0, y0, x1, y1;

				cin >> x0 >> y0 >> x1 >> y1;

				if(x0 > x1) swap(x0, x1);
				if(y0 > y1) swap(y0, y1);

				cout << p * bit.query(x0, y0, x1, y1) << '\n';
			}
		}

		cout << '\n';
	}
	return 0;
}